src/placement.c

```diff
@@ -0,0 +1,552 @@
+			r = m;
+		} else {
+			l = m + 1;
+		}
+	}
+
+	return r - 1;
+}
+
+/*
+ * Construct an overlap bitmap for the irregular grid, computed by
+ * build_grid(), that spans view->output. The overlap bitmap maps
+ * each interval to the number of views on the output (excluding *view)
+ * that overlap that interval.
+ */
+static void
+build_overlap(struct overlap_bitmap *bmp, struct view *view)
+{
+	assert(bmp);
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	if (bmp->nr_rows < 1 || bmp->nr_cols < 1) {
+		return;
+	}
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return;
+	}
+
+	struct view *v;
+	for_each_view(v, &server->views, LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		if (v == view || v->output != output) {
+			continue;
+		}
+
+		/* Find boundaries of the window */
+		struct border margin = ssd_get_margin(v->ssd);
+		int lx = v->pending.x - margin.left;
+		int ly = v->pending.y - margin.top;
+		int hx = v->pending.x + v->pending.width + margin.right;
+		int hy = v->pending.y + v->pending.height + margin.bottom;
+
+		/*
+		 * Find the first and last row and column intervals spanned by
+		 * this view. We want the left and top edges to fall in a
+		 * half-open interval [low, high) but the right and bottom
+		 * edges to fall in a half-open interval (low, high] to ensure
+		 * that the results do not include intervals adjacent to the
+		 * view. View edges are guaranteed by construction to fall
+		 * exactly on the grid points, so we perturb the left and top
+		 * edges by +0.5 units, and the right and bottom edges by -0.5
+		 * units, to ensure that we are always searching in the
+		 * interior of an interval.
+		 */
+
+		/* First row and column overlapping the view */
+		int fc = find_interval(bmp->cols, bmp->nr_cols, lx + 0.5);
+		int fr = find_interval(bmp->rows, bmp->nr_rows, ly + 0.5);
+
+		/* Clip first row/column to start of usable grid */
+		fc = MAX(fc, 0);
+		fr = MAX(fr, 0);
+
+		/* Last row and column overlapping the view */
+		int lc = find_interval(bmp->cols, bmp->nr_cols, hx - 0.5);
+		int lr = find_interval(bmp->rows, bmp->nr_rows, hy - 0.5);
+
+		/*
+		 * Increment the last indices to convert them to strict upper
+		 * bounds, then clip them to the limits of the usable grid.
+		 */
+		lc = MIN(bmp->nr_cols - 1, lc + 1);
+		lr = MIN(bmp->nr_rows - 1, lr + 1);
+
+		/*
+		 * Every interval in the region [fr, lr) x [fc, lc) is
+		 * completely covered by the view. Increment the overlap
+		 * counters these intervals to account for the view.
+		 */
+		for (int i = fr; i < lr; ++i) {
+			for (int j = fc; j < lc; ++j) {
+				overlap_bitmap_index(bmp, i, j) += 1;
+			}
+		}
+	}
+}
+
+/*
+ * Build an overlap bitmap for view->output, which subdivides view->output into
+ * a grid and then counts the number of windows that overlap each grid region.
+ */
+static struct overlap_bitmap
+build_bitmap(struct view *view)
+{
+	assert(view);
+
+	struct overlap_bitmap bitmap = { 0 };
```

Comment:
For this type of stuff, it's generally better to declare at the highest stack frame that the variable is used in (and easier for me to read).

Suggest we could get rid of this whole function and just add the following to where `build_bitmap()` is called from:

```
struct overlap_bitmap bitmap = { 0 };
build_grid(&bitmap, view);
build_overlap(&bitmap, view);
```

The `if (bitmap.nr_rows < 1...` check is done in `build_overlap()` anyway.